backend/attachRendererFiber.js

```diff
@@ -0,0 +1,252 @@
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @flow
+ */
+'use strict';
+
+import type {Hook, ReactRenderer, Helpers} from './types';
+var getDataFiber = require('./getDataFiber');
+
+function attachRendererFiber(hook: Hook, rid: string, renderer: ReactRenderer): Helpers {
+  // The naming is confusing.
+  // They deal with opaque nodes (fibers), not elements.
+  function getNativeFromReactElement(fiber) {
+    try {
+      return renderer.findHostInstanceByFiber(fiber);
+    } catch (err) {
+      // The fiber might have unmounted by now.
+      return null;
+    }
+  }
+  function getReactElementFromNative(node) {
+    const fiber = renderer.findFiberByHostInstance(node);
+    return fiber;
+  }
+
+  // This is a slightly annoying indirection.
+  // It is currently necessary because DevTools wants
+  // to use unique objects as keys for instances.
+  // However fibers have two versions.
+  // We use this set to remember first encountered fiber for
+  // each conceptual instance.
+  const opaqueNodes = new Set();
+  function getOpaqueNode(fiber) {
+    if (opaqueNodes.has(fiber)) {
+      return fiber;
+    }
+    const {alternate} = fiber;
+    if (alternate != null && opaqueNodes.has(alternate)) {
+      return alternate;
+    }
+    opaqueNodes.add(fiber);
+    return fiber;
+  }
+
+  function haveChildrenChanged(prevChildren, nextChildren) {
+    if (!Array.isArray(prevChildren) || !Array.isArray(nextChildren)) {
+      return prevChildren !== nextChildren;
+    }
+    if (prevChildren.length !== nextChildren.length) {
+      return true;
+    }
+    for (let i = 0; i < prevChildren.length; i++) {
+      if (prevChildren[i] !== nextChildren[i]) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  function hasDataChanged(prevData, nextData) {
+    return (
+      prevData.ref !== nextData.ref ||
+      prevData.source !== nextData.source ||
+      prevData.props !== nextData.props ||
+      prevData.state !== nextData.state ||
+      prevData.context !== nextData.context ||
+      prevData.text !== nextData.text ||
+      haveChildrenChanged(prevData.children, nextData.children)
+    );
+  }
+
+  let pendingEvents = [];
+
+  function flushPendingEvents() {
+    const events = pendingEvents;
+    pendingEvents = [];
+    for (let i = 0; i < events.length; i++) {
+      const event = events[i];
+      hook.emit(event._event, event);
+    }
+  }
+
+  function enqueueMount(fiber) {
+    pendingEvents.push({
+      // TODO: the naming is confusing. `element` is *not* a React element. It is an opaque ID.
+      element: getOpaqueNode(fiber),
+      data: getDataFiber(fiber, getOpaqueNode),
+      renderer: rid,
+      _event: 'mount',
+    });
+
+    const isRoot = fiber.tag === 3;
```

Comment:
nit Because of references like this and the ones in `getDataFiber` I wonder if it wouldn't be clearer to define a duplicate `ReactTypeOfWork` const in this project. At least this way if a value ever changes it's clear where all of the mirrored values are (and we aren't just greping for numbers).